pulp-dev/pulp_dev/releng_utils.py

```diff
@@ -0,0 +1,843 @@
+        version component is "2", the patch version component is "3". The release component is
+        "0.1.beta", made up of release major component "0", release minor component "1", and
+        release stage "beta".
+
+        Unless stage is True or next_stage is set, the release stage will not advance through the
+        "alpha" -> "beta" -> "rc" -> "ga" cycle. Before GA, the major release component will
+        always be 0, and the minor release component will be incremented. Upon moving to GA, the
+        release minor component and stage will be ignored, and only the release major component
+        will be incremented.
+
+        Incrementing the epoch is not supported.
+
+        :param major: Increment the major version component
+        :param minor: Increment the minor release component
+        :param patch: Increment the patch release component
+        :param release: Increment to next release stage with appropriate release version
+        :param stage: Increment to next stage
+        :param next_stage: Force stage in incremented version to this stage (release minor
+                           component will be incremented)
+
+        """
+        copy = self.copy()
+
+        # If a next stage was passed in, use it. If not, calculate the next stage if "stage" is
+        # going to be incremented, otherwise just use the current stage.
+        # next_stage will remain None if the next stage is GA, or release is True
+        if next_stage is None:
+            if stage:
+                next_stage = copy.next_stage()
+            elif release:
+                next_stage = None
+            else:
+                next_stage = copy.r_stage
+
+        # fun failsafe: if next_stage is still None at this point, release gets forced to
+        # True, since that's what a next_stage of None means.
+        if next_stage is None:
+            release = True
+
+        if any((major, minor, patch)):
+            # anything updating the version gets a new release value: 0.1.<stage>
+            # unless release is True, then update the release major and reset release minor
+            if release:
+                copy.r_major += 1
+                copy.r_minor = 0
+            else:
+                copy.r_major = 0
+                copy.r_minor = 1
+            copy.r_stage = next_stage
+
+        if major:
+            # 1.2.3 becomes 2.0.0
+            copy.v_major += 1
+            copy.v_minor = copy.v_patch = 0
+        elif minor:
+            # 1.2.3 becomes 1.3.0
+            copy.v_minor += 1
+            copy.v_patch = 0
+        elif patch:
+            # 1.2.3 becomes 1.2.4
+            copy.v_patch += 1
+        elif next_stage:
+            # always increment the minor release if advancing to a non-release stage
+            copy.r_minor += 1
+            if stage:
+                copy.r_stage = next_stage
+
+        return copy
+
+
+# py3: Remove this, use tempfile.TemporaryDirectory
```

Comment:
The only reason I got py3 involved was part of an effort to get the "editable: True" to actually install pulp-dev as editable in the ansible tasks, which means I can just clean this up and go back to all-py3-all-the-time in this PR (which I'll do)